src/memory-management/comparison.md

```diff
@@ -24,12 +24,12 @@ Here is a rough comparison of the memory management techniques.
   * Double-frees.
   * Memory leaks.
 * Automatic like Java:
+  * Runtime overhead.
   * Garbage collection pauses.
   * Destructor delays.
 * Scope-based like C++:
-  * Complex, opt-in by programmer (on C++).
-  * Circular references can lead to memory leaks
-  * Potential runtime overhead
+  * Complex, opt-in by programmer.
```

Comment:
This is a very c++-centric point, and not true for other languages

Reply:
You mentioned ARC in Swift at some point, so I guess you might be thinking of this? Similar to the automatic reference counting done by Python, I'm not sure if ARC falls under the heading of scope-based memory management...

To be more precise: using `std::unique_ptr` in C++ comes with _zero overhead_ over using raw pointers. The compiler will insert precisely the necessary instructions to determine when to free the pointer, even if there are early returns etc. You can implement this in any language that has a concept of a destructor.

I'm not sure how to classify reference counting (manuel in C, `std::shared_ptr` in C++, ARC in Swift, and Python) in this sense.

Reply:
Yes I'm thinking of `std::shared_ptr`, I don't think you can write complete software with `unique_ptr` alone, can you?

Reply:
Ok read up on what scope-based memory management actually is, it's pretty much just stack allocated variables in C, and a bit more beef in C++? In this case the main drawback should be that it only works for simple non-dynamic programs, right?

Reply:
@domenukk Reading your comment I realize that the course equates automatic memory management to garbage collection (in summary.md, automatic memory management is called garbage collection).

In my mental model, automatic memory management could be thought of as an interface for the programmer that allows them to not care about the memory lifecycles. While GC vs RC is an implementation detail.

Scope based forces the programmer to care about the lifecycles through move and reference semantics while abstracting away the actual allocations and destructions. So the interface is different for the programmer.

Maybe the class doesn't need to mention Garbage Collection or reference counting at all and just focus on automatic vs manual vs RAII?